R/reactive-cached.R

```diff
@@ -0,0 +1,384 @@
+#'     cacheKeyExpr = list(input$x, input$y),
+#'     valueExpr = { input$x + input$y }
+#'   )
+#'   ```
+#'
+#' The first time someone calls `r()`, it executes both `cacheKeyExpr` and
+#' `valueExpr`. If someone calls `r()` again, then it does not need to
+#' re-execute `cacheKeyExpr`, because that `expr` has not been invalidated
+#' via a change to `input$x` or `input$y`; it simply returns the previous value.
+#' However, if `input$x` or `input$y` changes, then the reactive expression will
+#' be invalidated, and the next time that someone calls `r()`, `cacheKeyExpr`
+#' will need to be re-executed.
+#'
+#'
+#' @section Event expressions and reactivity:
+#'
+#'   Typically, the `cacheKeyExpr` is reactive, and the `valueExpr` is not (it
+#'   is run within [isolate()]). However, there are times when this isn't ideal:
+#'   perhaps you have [sliderInput]s `x` and `y`, but you don't want the
+#'   computation to occur until the user sets both `x` and `y`, and then clicks
+#'   on an [actionButton] named `go`.
+#'
+#'   The value of `input$go` shouldn't be included in the cache key, because its
+#'   value is not relevant to the calculation involving `input$x` and `input$y`.
+#'   You also don't want to take a reactive dependency on `input$x` and
+#'   `input$y`, because then any changes to those values would cause this cached
+#'   reactive to do the computation or fetch the value from the cache and return
+#'   it.
+#'
+#'   In short, in this case, you want to take a reactive dependency on
+#'   `input$go` and not use its value for the cache key, and you want to **not**
+#'   take a reactive dependency on `input$x` and `input$y` but use those values
+#'   for the cache key.
+#'
+#'   This can be done by using `eventExpr`, similar to in the [eventReactive()]
+#'   function. If a non-NULL `eventExpr` is provided to `cachedReactive()`, then
+#'   it will be used for the reactive dependencies but its value will be
+#'   ignored, and the `cacheKeyExpr` will be used for its value, but it will be
+#'   executed in an [isolate()], so it will not be used for its reactive
+#'   dependencies.
+#'
+#'   In the example described above, you would use something like:
+#'
+#'   ```
+#'   r <- cachedReactive(
+#'     eventExpr = input$go,
+#'     cacheKeyExpr = list(input$x, input$y),
+#'     valueExpr = { input$x + input$y }
+#'   )
+#'   ```
+#'
+#' @section Async with cached reactives:
+#'
+#'   With a cached reactive expression, they key and/or value expression can be
+#'   _asynchronous_. In other words, they can be promises -- not regular R
+#'   promises, but rather objects provided by the promises package, which are
+#'   similar to promises in JavaScript. (See [promises::promise()] for more
+#'   information.) You can also use [future::future()] objects to run code in
+#'   a separate process or even on a remote machine.
+#'
+#'   If `valueExpr` is a promise, then anything that consumes the
+#'   `cachedReactive` must expect it to return a promise.
```

Comment:
Yup, a `reactive` returns a promise -- that's the only way it could work.

```R
library(shiny)
library(promises)

x <- reactive({ promise_resolve("hello") })

observe({
  x()$then(function(value) {
    message(value)
  })
})

shiny:::flushReact()
#> hello
```

What you're proposing is a much, much more difficult thing to do, although there is this: https://github.com/r-lib/coro